tortoisebeacon/weakcoin/weak_coin.go

```diff
@@ -4,290 +4,303 @@ import (
+		wc.smallestProposal = proposal
 	}
 }
 
-func (wc *weakCoin) OnRoundFinished(epoch types.EpochID, round types.RoundID) {
-	wc.Log.With().Info("Round finished",
-		log.Uint64("epoch_id", uint64(epoch)),
-		log.Uint64("round_id", uint64(round)))
-
-	wc.activeRoundsMu.Lock()
-	defer wc.activeRoundsMu.Unlock()
-
-	pair := epochRoundPair{
-		EpochID: epoch,
-		Round:   round,
-	}
-
-	if _, ok := wc.activeRounds[pair]; ok {
-		delete(wc.activeRounds, pair)
-		wc.calculateWeakCoin(epoch, round)
-	}
+func (wc *WeakCoin) exceedsThreshold(proposal []byte) bool {
+	return bytes.Compare(proposal, wc.config.Threshold) == 1
 }
 
-func (wc *weakCoin) calculateWeakCoin(epoch types.EpochID, round types.RoundID) {
-	pair := epochRoundPair{
-		EpochID: epoch,
-		Round:   round,
-	}
-
-	wc.proposalsMu.RLock()
-	defer wc.proposalsMu.RUnlock()
-
-	wc.weakCoinsMu.Lock()
-	defer wc.weakCoinsMu.Unlock()
-
-	proposals := wc.proposals[pair]
-
-	if len(proposals) == 0 {
-		wc.Log.With().Warning("No weak coin proposals were received",
-			log.Uint64("epoch_id", uint64(epoch)),
-			log.Uint64("round_id", uint64(round)))
-
-		wc.weakCoins[pair] = false
-
-		return
+func (wc *WeakCoin) encodeProposal(epoch types.EpochID, round types.RoundID, unit uint64) []byte {
+	proposal := bytes.Buffer{}
+	proposal.WriteString(wc.config.VRFPrefix)
+	if _, err := proposal.Write(epoch.ToBytes()); err != nil {
+		wc.logger.Panic("can't write epoch to a buffer", log.Err(err))
 	}
-
-	smallest := new(big.Int).SetBytes(proposals[0])
-	stringProposals := []string{smallest.String()}
-
-	for i := 1; i < len(proposals); i++ {
-		current := new(big.Int).SetBytes(proposals[i])
-		stringProposals = append(stringProposals, current.String())
-
-		if current.Cmp(smallest) == -1 {
-			smallest = current
-		}
+	roundBuf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(roundBuf, uint64(round))
+	if _, err := proposal.Write(roundBuf); err != nil {
+		wc.logger.Panic("can't write uint64 to a buffer", log.Err(err))
 	}
-
-	coinBit := new(big.Int).And(smallest, big.NewInt(1))
-
-	weakCoinValue := coinBitToBool(coinBit)
-	wc.weakCoins[pair] = weakCoinValue
-
-	wc.Log.With().Info("Weak coin calculation input",
-		log.Uint64("epoch_id", uint64(epoch)),
-		log.Uint64("round_id", uint64(round)),
-		log.String("smallest", smallest.String()),
-		log.String("coin_bit", coinBit.String()),
-		log.String("proposals", strings.Join(stringProposals, ", ")))
-
-	wc.Log.With().Info("Calculated weak coin",
-		log.Uint64("epoch_id", uint64(epoch)),
-		log.Uint64("round_id", uint64(round)),
-		log.Bool("weak_coin", weakCoinValue))
-}
-
-func coinBitToBool(coinBit *big.Int) bool {
-	switch coinBit.Int64() {
-	case 0:
-		return false
-	case 1:
-		return true
-	default:
-		panic("bad input")
+	buf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(buf, unit)
+	if _, err := proposal.Write(buf); err != nil {
+		wc.logger.Panic("can't write uint16 to a buffer", log.Err(err))
```

Comment:
I guess you meant `uint64` here
```suggestion
		wc.logger.Panic("can't write uint64 to a buffer", log.Err(err))
```